File: 2020/fsharp/Day01.fs
namespace AdventOfCode.FSharp.Y2020

module Day01 =
    open AdventOfCode.FSharp.Util

    let parse input = input |> splitLine |> Array.map int

    let part1 input =
        let x = input |> parse

        Array.allPairs x x
        |> Seq.pick (fun (a, b) -> if a <> b && (a + b) = 2020 then Some(a * b) else None)
        |> string

    let part2 input =
        let x = input |> parse

        Array.allPairs x x
        |> Seq.allPairs x
        |> Seq.pick (fun (a, (b, c)) ->
            if a <> b && a <> c && (a + b + c) = 2020 then
                Some(a * b * c)
            else
                None)
        |> string

    let run (input: byte array) output =
        let textInput = text input
        part1 textInput |> output 1
        part2 textInput |> output 2

File: 2020/csharp/Day01.cs
using System.Text;

namespace AdventOfCode.CSharp.Y2020;

public static class Day01
{
    public static void Run(byte[] input, Action<int, string> output)
    {
        var Input = Encoding.UTF8
            .GetString(input)
            .Split('\n', StringSplitOptions.RemoveEmptyEntries);

        int Part1()
        {
            var numbers = from line in Input
                          where !string.IsNullOrWhiteSpace(line)
                          select int.Parse(line);

            var na = numbers.ToArray();

            var find =
                from x in na
                from y in na
                where x != y && (x + y) == 2020
                select x * y;


            return find.First();
        }

        int Part2()
        {
            var numbers = from line in Input
                          where !string.IsNullOrWhiteSpace(line)
                          select int.Parse(line);

            var na = numbers.ToArray();

            var find =
                from x in na
                from y in na
                from z in na
                where x != y && y != z && (x + y + z) == 2020
                select x * y * z;

            return find.First();
        }

        output(1, Part1().ToString());
        output(2, Part2().ToString());
    }
}